[Sheyuan Yeo] iP

Piyotato opened the pull request:
### Duke
> Don't remember things, write them down. - Me

Don't forget things ever again because you didn't write them down. Duke provides a place for you to write down things you cannot remember. It is

1. cool
- super cool
- mega cool
- ~~mega~~ **_ultra cool_** 😉

and even saves to file!

has [amazing](https://c.tenor.com/DTxBtpexadYAAAAC/wojak-rage-wojak.gif) code including

```java
this.parser = new Parser(new ArrayList<>(Arrays.asList(
                new Exit("bye", this.ui),
                new AddTaskCommand<>("todo", taskList, null, ToDo::new),
                new AddTaskCommand<>("deadline", taskList, " /by ", Deadline::new),
                new AddTaskCommand<>("event", taskList, " /at ", Event::new),
                new ListCommand("list", taskList),
                new MarkCommand("mark", taskList, true),
                new MarkCommand("unmark", taskList, false),
                new DeleteTaskCommand("delete", taskList),
                new FindCommand("find", taskList)
        )));
```

if anything wrong, try
- [ ] close and open again
- [ ] shut down and try again
- [ ] cry

Commit:
Add Gradle support

Piyotato committed:
Add skeletal version of Duke

- Add user greeting message
- Add user input echoing
- Add exit command with exit message

Piyotato committed:
Implement list system

- Input is now stored into a list instead simply being repeated
- Add "list" command which prints the items in the list in an enumerated manner
- Move all commands into constant strings for ease of access
- Add documentation for methods

Piyotato committed:
Implement task list mark/unmark

- Add mark/unmark commands
- Add Task class that represents a task on the list

Piyotato committed:
Implement ToDo, Deadline, Event classes

- Make Task no longer a private inner class in Duke
- Add derived classes ToDo, Deadline, Event that extend Task
- Add todo, deadline, event commands
- Add some command parsing safety

Piyotato committed:
Add Automated Text UI Testing

Piyotato committed:
Catch Exceptions

- Catch empty todo/deadline/event. Currently does not treat whitespaces as empty.
- Catch invalid commands
- Update Automated Text UI testing

Piyotato committed:
Support Deletion

- Support deletion using "delete" command
- Fix off-by-1 error when detecting index out of range for mark/unmark
- Add message of encouragement when task list is empty
- Update Automated Text UI Testing

Piyotato committed:
Refactor Duke Structure

- Organize into packages (duke.core, duke.command, duke.task)
- Make Duke more OOP-driven by abstracting command, parser, task
- Update test cases
- Update runtest.sh

Piyotato committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into week3

Piyotato committed:
Add Saving System

- Duke now saves the contents of TaskList into a file (duke.txt)
- Add serialize/deserialize methods for Task
- Duke now loads from this task list when starting up

Piyotato committed:
Add date parsing to Deadline and Event

- Make Deadline and Event use dates instead of arbitrary strings as the deadline/timing
- Add protected getFormattedDetails() method in Task for standardizing how this date is displayed
- Update input tests to account for this change

Piyotato committed:
Merge branch 'branch-level-8'

Piyotato committed:
Fix merge issues with deserialization

- Account for date when deserializing task

Piyotato committed:
Merge branch 'week3'

Piyotato committed:
Fix typo

Piyotato committed:
Add JUnit Tests

- Add JUnit tests for various Task methods for each of the task types.
- Remove some unused constants.
- Fix some style issues.

Piyotato committed:
Delete duke.txt

Piyotato committed:
Add duke.txt to .gitignore

Piyotato committed:
Support building in gradle, add duke.txt to .gitignore

- Fix mainClassName in build.gradle
- Support building of Duke as a task in gradle.

Piyotato committed:
Merge branch 'branch-a-jar'

Piyotato committed:
Add JavaDocs

Piyotato committed:
Add searching

- Add a find command that searches the task list for tasks with matching titles and prints them out.

Piyotato committed:
Fix JavaDocs style

Piyotato committed:
Add checkstyle, fix checkstyle errors

nehcuy commented on src/main/java/duke/commands/Exit.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * A command that exits a UI.
+ */
+public class Exit extends Command {
```
Is it better if this class was named ExitCommand for consistency of the Command subclasses?

nehcuy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,131 @@
+    private static final String serializerSeparator = "---";
+    protected String text = "";
+    protected LocalDate details;
+    protected boolean complete = false;
```
Perhaps `isComplete` would be a better name for a boolean?

nehcuy commented on src/main/java/duke/core/Ui.java:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * A line that is printed to show text boundaries.
+     */
+    private static final String line = "____________________________________________________________";
```
Since these are your constants, it would be better to capitalise the whole name. Eg. `WELCOME_MESSAGE`, `INDENTATION`, `LINE`.

nehcuy commented on src/main/java/duke/core/Ui.java:
```diff
@@ -0,0 +1,87 @@
+
+    public boolean isExit() {
+        return isExit;
+    }
```
I think it's better to have consistency, so you could either do `setExitStatus()` and `getExitStatus()`, or simply rename the `isExit()` getter to be `getIsExit()` 😄

nehcuy commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,131 @@
+ * A class that represents a task in the task list.
+ */
+public abstract class Task {
+    private static final String serializerSeparator = "---";
```
From what I understand, this is a constant, therefore it might be better to make the variable declaration capitalised. i.e. `SERIALIZER_SEPARATOR`

nehcuy commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,96 @@
+     *
+     * @return String with all the tasks separated by newlines.
+     */
+    public String listTasks() {
```
Not a big issue, but even though the word _list_ in `listTask()` and class `TaskList` have different meanings, it sounds a little confusing. Therefore, it might be better to use `printListOfTasks()` instead.

nehcuy commented on src/main/java/duke/commands/MarkCommand.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * The completion status that this command will mark a task as.
+     */
+    private final boolean statusToMark;
```
I feel like a more suitable boolean variable name would be `isMarked`

nehcuy commented on src/main/java/duke/commands/MarkCommand.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * The text that is displayed when a task is unmarked.
+     */
+    private static final String unmarkedText = "Noted. I've unchecked this task:";
```
Constant names should probably be all capitalised: `MARKED_TEXT` and `UNMARKED_TEXT`

nehcuy commented on src/main/java/duke/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Text to be displayed when this task is successfully executed.
+     */
+    private static final String taskAddedText = "Got it. I've added this task:";
```
I think you should capitalise constant variable names.

nehcuy commented on src/main/java/duke/commands/Exit.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * The exit message used when the user issues the exit command.
+     */
+    private static final String exitMessage = "Bye. Hope to see you again soon!";
```
Remember to capitalise constant variables.

nehcuy commented on src/main/java/duke/commands/FindCommand.java:
```diff
@@ -0,0 +1,39 @@
+public class FindCommand extends TaskListCommand {
+
+    private static final String successText = "Here are the matching tasks in your list:";
+    private static final String failureText = "No matching tasks found.";
```
Remember to capitalise constant variable names to `SUCCESS_TEXT` and `FAILURE_TEXT`

nehcuy commented on src/main/java/duke/commands/ListCommand.java:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * Text to display before enumerating the tasks in the list.
+     */
+    private static final String tasksText = "Here are the tasks in your list:";
```
Remember to capitalise constant variable names. 😄

nehcuy reviewed (commented):
LGTM after correcting capitalisation of constant variable declarations. Cool usage of generic types! 😃

Piyotato committed:
Fix constant naming, adjust variable names, adjust ExitCommand name

ruihan00 commented on src/main/java/duke/core/Duke.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    public static void main(String[] args) {
+        new Duke("duke.txt").run();
```
very neat

ruihan00 commented on src/main/java/duke/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,102 @@
+     */
+    public AddTaskCommand(String invoker, TaskList taskList, String argumentDelimiter,
+                          Supplier<? extends T> taskFactory) {
+        super(invoker, taskList);
```
cool!

ruihan00 commented on src/main/java/duke/core/Duke.java:
```diff
@@ -0,0 +1,108 @@
+     */
+    public Duke(String fileName) {
+        this.parser = new Parser(new ArrayList<>(Arrays.asList(
+                new ExitCommand("bye", this.ui),
+                new AddTaskCommand<>("todo", taskList, null, ToDo::new),
+                new AddTaskCommand<>("deadline", taskList, " /by ", Deadline::new),
```
very neat

ruihan00 commented on src/main/java/duke/core/Parser.java:
```diff
@@ -0,0 +1,50 @@
+     * @return Command output.
+     * @throws DukeException If command is invalid or if any execution is erroneous.
+     */
+    public String parseInput(String input) throws DukeException {
```
interesting

ruihan00 reviewed (commented):
nice code

Piyotato committed:
Use JavaFX to create GUI for Duke

- Remove text-ui-test since there is no more text UI
- Remove Ui class as display is now handled by JavaFX
- Adjust ExitCommand to exit the Main application
- Adjust Duke to work with JavaFX application instead of Ui

Piyotato committed:
Reintroduce Welcome Message

Currently, the welcome message is mistakenly removed when moving to JavaFX UI.

Without it, the user starts up to an empty window, which makes Duke seem incomplete.

Reintroducing it in the MainWindow makes sense because it is the new equivalent to the old Ui class, where the message used to be displayed.

Let's
- reintroduce the welcome message in MainWindow
- add a bit of personality to it

Piyotato committed:
Merge pull request #1 from Piyotato/bug-welcome-message

Reintroduce Welcome Message

Piyotato committed:
Add assertions

- Add assertion to document the assumption that the invoker in `Command` is not empty.
- Add assertion to document the assumption that argument delimiters are not blank in `AddTaskCommand`

Piyotato committed:
Merge pull request #2 from Piyotato/implement-a-assertions

Add assertions

Piyotato committed:
Improve code quality

- Remove debug print statement in `Task`
- Fix complicated expression in `Task`
- Add newlines in `Duke` for readability
- Make happy path more prominent in `TaskListIndexCommand`
- Make happy path more prominent in `AddTaskCommand`

Piyotato committed:
Merge pull request #3 from Piyotato/implement-a-codequality

Improve code quality

Piyotato committed:
Add continuous integration

Piyotato committed:
Remove text-ui checks from continuous integration

Piyotato committed:
Fix locale issues with unit tests discovered in CI

Piyotato committed:
Remove unused import

Piyotato committed:
Fix checkstyle error

Piyotato committed:
Merge pull request #4 from Piyotato/implement-a-ci

Add continuous integration

Piyotato committed:
Add duplication detection

- Task can no longer be added if a duplicate exists in the list
- Rework logic to make use of lazy evaluation in `Task::equals`

Piyotato committed:
Merge pull request #5 from Piyotato/implement-c-detectduplicates

Add duplication detection

Piyotato committed:
Fix resizing, font issues

- Make resizing work with Duke
- Change font to Arial to avoid display issues in other OSes

Piyotato committed:
Merge pull request #7 from Piyotato/gui

Fix resizing, font issues

Piyotato committed:
Add user guide

Piyotato committed:
Merge pull request #8 from Piyotato/user-guide

Add user guide

Piyotato committed:
Fix FXML version

Piyotato committed:
Merge pull request #9 from Piyotato/gui

Fix FXML version
